### ISSUE TITLE
Should Semiregular impose copy-indenpendence?

### ISSUE THREAD

**Issue by tomaszkam** [ericniebler/stl2#38]
I am creating a separate topic for the issue that was mentioned during discussion about `reference_wrapper<I>` in the topic [https://github.com/ericniebler/stl2/issues/23](algorithm return types should maybe drop pair and tuple in favor of structs with more meaningful field names #23).

The question is whether the `Semiregular` concept requires the copy independence, i.e. guarantee that the modification of the copy will not change it source. This may be rephrased into the question if `SemiRegular` types should have value semantics or they are allowed to expose reference semantics.

The non normative node placed in the current _Range TS_ on `Semiregular`:
> The `Semiregular` concept is modeled by types that have behave similarly to built-in types like
int, except that they may not be comparable with `==`. 
Suggest that the types that meets this concept should provide value like semantics (like `int`).

From the other side requiring copy independence from `Semiregular` would break input/output iterators (as pointed by _CaseyCarter_). However I think that the problem should be solved by removing `Semiregular` concept from `WeaklyImplementable`. I believe that such types should not be required to be `CopyConstructible`, because copying the value of the iterator to save current position is undefined for this type of objects, so no algorithm that accepts  these type of iterator needs to create a copy of iterator. To pass is argument to function/return, move operation would be sufficient,

Furthermore copy constructor of `std::istream_iterator` behaves similarly to `std::auto_ptr`- after copy only one of the objects may be used. I think that similar shift like the one in case of std::unique_ptr could be also applied here, which will clearly suggest that now move-from (copied-from before) iterator is not valid.

In conclusion I am suggesting to add copy independence requirements to `SemiRegular` as semantics requirement. And reduce requirement for  `WeaklyImplementable` to `Moveable` and `DefaultConstrutible`.

**Comment by ericniebler** [comment]
It has been suggested before that Input iterators should be move-only. I think it would break too much code.

**Comment by tomaszkam** [comment]
> It has been suggested before that Input iterators should be move-only. I think it would break too much code.

I am saying that they should not require to be copyable, so new input iterators can be move-only. That does not prohibit making them copy constructible.

**Comment by ericniebler** [comment]
I understand your suggestion. The situation is different than `auto_ptr`. There, we could deprecate `auto_ptr` and provide a better component. Here, there will be lots of code in the wild that expects copyability from Input iterators, and there is nothing we can deprecate and nothing we can give them as a replacement. Their code is just broken, and in a major way. Iterators are copied everywhere. The other breakages you mention are fairly minor (I hope). When considering breaking changes, a cost-benefit analysis must be done.

**Comment by tomaszkam** [comment]
> There, we could deprecate auto_ptr and provide a better component. Here, there will be lots of code in the wild that expects copyability from Input iterators, and there is nothing we can deprecate and nothing we can give them as a replacement.

While `make_common_iterator` (or any other wrapping mechanism for STL1) could not provide a wrapper that will implement copy in terms of move for stored iterator like for any other missing operation. 

We decide that we could break backward compatibility to fix bugs and we should be consequent with our decisions. After STL2 will be done the new code will be written using it for years, and if we leave "bugs" to keep backward compatibility, there will be no way to fix it.

**Comment by ericniebler** [comment]
> After STL2 will be done the new code will be written using it for years, and if we leave "bugs" to keep backward compatibility, there will be no way to fix it.

OTOH, if we make migration too painful, and make usage awkward by forcing users to move iterators and wrap them in copy-as-move helpers, will people use STL2 at all?

**Comment by tomaszkam** [comment]
> OTOH, if we make migration too painful, and make usage awkward by forcing users to move iterators and wrap them in copy-as-move helpers, will people use STL2 at all?

Let me describe my view. We do not touch existing input iterators declared in the standard at all, except maybe deprecating it (if we provide new component). That means that the code written before will compile, as the copy operation of input iterator used in the written user-defined algorithms and functions will still work (as they necessarily use STL1 `InputIterators`). For the new STL2 algorithms. the requirements are weaker so they will still compare. This does not break existing code in contrast to return type changes or requiring symmetric call operator (both justified).

So, what happens if we start to use new STL2 InputIterators. This is not matter of backward compatibility, because if we have new iterators declared, then we necessarily write new code. In that case the only problem is to use old user-define algorithm on the new iterators is to use `make_bounded_range` that is needed anyway, because of other needs.

I think that there is nothing awkward in use of the move to pass iterators. Firstly if we use range algorithms, the `begin` and `end` will be called inside of the library and they will return rvalues, i.e. non move is necessary in basic case. The only situations when move is explicitly required is when we store iterator. And I think that use of the explicit `move` will increase readableness of the code and reduce bugs caused _ackward_ copy semantic of existing iterators, by reducing it to common case of use of objects in moved from state. This is exactly the same as in `auto_ptr`, we may say that required use of `move` of `unique_ptr` is ackward and people will not use it. The copy operation that does not makes copies are awkward!

**Comment by asutton** [comment]
> OTOH, if we make migration too painful, and make usage awkward by forcing
> users to move iterators and wrap them in copy-as-move helpers, will people
> use STL2 at all?
>
> Let me describe my view.

<snip>

This needs a full paper that lays out motivation, design, and impact
of the change. This isn't a change that I would feel comfortable
presenting to L[E]WG in the next meeting without their input.

Andrew

**Comment by ericniebler** [comment]
@tomaszkam Do you feel strongly enough about this to write a paper?

**Comment by tomaszkam** [comment]
I think I have again tangled two topics into single issue, and I would good to provide same clarification.

The main issue that I was trying to emphasize that currently the `Semiregular` concept is used in two different places:
1. In case of the same algorithms copy-independence is required from the `ValueType<I>`, which is required to be only `Semiregular`. For example the `sort` implementation creates pivot by making copy of the element of range and the implementation assumes that changes done to range elements will not affect value of the pivot.
2. In case of `WeaklyIncrementable` types we explicitly allow types that have reference copy semantics, i.e. does not fulfill copy-independence requirements.
As this two set of the requirements are contradictory, we cannot use the same concept in both of the cases, because this will cause one of the case to be over or under constrained. We need to have separate set of the constrains for this two cases and it need to be fixed in current wording

My suggestion was to require copy-independence for `Semiregular` concept and create a new concept for the input iterators. This is based on the non-normative note placed in the current wording that suggest that semiregular concept are intended to represent types similar to build-in types and I believe that includes value-semantics. However this solution still does not require removing `Copyable` requirement for input iterators.

The second presented suggestion was to remove `Copyable` requirement for input iterators. This should be considered as the suggestion not as the issue report. 

> @tomaszkam Do you feel strongly enough about this to write a paper?

I assume that you are asking about the removing the capability requirement for input iterator, not about the problem with `Semiregular`. In that case I am willing to try to write a paper. 

However my current experience (a few papers with small library additions) does not me confident that I will able to handle it correctly, so I you feel more convinced to my view, I will gladly accept both your help or taking care about the issue.

**Comment by CaseyCarter** [comment]
There's some confusion here - and I'm afraid my comment that input iterators cannot support copy independence is the source. EoP's Iterator model refines Regular, despite the fact that the successor operation is not required to be equality-preserving (pg. 91). Copy independence requires that copies are distinct: changing the value of an object should not change the value of a copy of that object. The confusion here is that we are conflating (a) the value of an iterator with the value the iterator denotes, and (b) the notions of modifying an object's value with invalidating an object's value.

Pointer types are regular and have copy independence despite the fact that two copies of the same value point at the same object because the value of the denoted object is not part of the value of the pointer. After assigning `p = q` for pointers `p` and `q`, we know that `p == q` holds regardless of modifications to the object they denote. Reference types - and similarly `reference_wrapper` and other "smart references" - notably do NOT share this property. Since equality on reference types is defined as equality of the referents, a reference type's value includes the value of the object it denotes. Reference types by definition do not have copy independence.

Pointer types also have the notion that a value which is well-defined for an operation at some point in the program may not be well-defined at another point. E.g, after executing `int *p = new int{67}, *q = p;` `q` has a value that is well-defined for dereference. But after executing `delete p;` the value of `q` no longer has that property; the Standard tells is that passing a pointer to a block of memory to a deallocation function invalidates the values of any pointers into that block. Note that validity is a property of values and operations, not a property of *objects*. The validity of an object's value may change at many points in a program without any modification to the object, e.g.:
```
struct A {};
void f() {
  struct A { int i; };  
  A* const p = new A{42};
  int* const q = &p->i;
  // p and q hold valid values and may be dereferenced
  for (auto i = 0; i < 42; ++i) {
    p->~A();
    // p and q hold invalid values
    ::new(p) A{i};
    // p and q are valid
  }
  delete p;
  // p and q are invalid
}
```

As iterators are an abstraction of pointer types and "smart references" (e.g `reference_wrapper`) an abstraction of reference types, these arguments for pointers and references generalize to iterators and smart references. The fact that applying `++` to an InputIterator object may invalidate its prior value is no different from the invalidation of pointer values caused by passing them to a deallocation function.

Input iterators model Semiregular, even if Semiregular requires copy independence. Smart references, and specifically `reference_wrapper`, can model Semiregular only in limited domains.

An aside on the idea of making InputIterator move-only: this would necessitate a move-only InputSentinel as well:
```
template <class I>
concept bool InputIterator = Movable<I> && ...
template <class S, class I>
concept bool InputSentinel = InputIterator<I> && Movable<S> && EqualityComparable<I, S>();
```
and make it necessary to return both from every function that does not traverse the entire sequence, since there would be no way for a caller to recover the value of the sentinel. i.e.:
```
template <InputIterator I, InputSentinel<I> S, EqualityComparable<ValueType<I>> T>
std::pair<I, S> find(I first, S last, const T& t);
```
`InputRange` will consequently become move-only as well. The same arguments about invalidation apply to output iterators, so `OutputIterator` and `OutputRange` (if/when we decide we want to support e.g. 4-legged copy) should also become move-only. I'm not arguing for or against, just pointing out that this is a sizable change.

**Comment by tomaszkam** [comment]
> Input iterators model `Semiregular`, even if `Semiregular` requires copy independence. Smart references, and specifically `reference_wrapper`, can model Semiregular only in limited domains.

What about the `iterator_reference_wrapper` proposed in the other thread? In that case changing a copy will change source (i.e. type is not copy-independent). Do we want to support such classes as to be used as `InputIterators`?

**Comment by CaseyCarter** [comment]
> What about the iterator_reference_wrapper proposed in the other thread? In that case changing a copy will change source (i.e. type is not copy-independent). Do we want to support such classes as to be used as InputIterators?

The question isn't "do we want to support it?" so much as "does such an iterator model InputIterator?" I'm reasonably certain the answer to that question is yes, if the domain is limited to wrapper objects that do not alias each other. If a function is only given access to non-aliased wrappers, there's no way for it to derive aliased wrappers and detect said aliasing without violating the InputIterator requirements.

**Comment by ericniebler** [comment]
I haven't been following this too closely, but I believe such a type would have limited utility. Since at best it could only model `InputIterator`, it cannot be used with any algorithms that require stronger iterators. It isn't a general solution to the problem of needing to return the end iterator position.

**Comment by CaseyCarter** [comment]
It wasn't intended to general, so much as a workaround to recover the end iterator from the algorithms in the set "Basically, if the whole point if the algorithm is to compute one value, I opted not to change the return type." A quick scan through the algorithm synopsis shows the wrapper would be sufficient for:

>all_of, any_of, none_of, count, count_if, equal, is_partitioned, includes, set_intersection, set_difference, min, max, minmax, lexicographical_compare

which all have input iterator overloads, but not for:

>is_permutation, is_sorted, binary_search, is_heap, min_element, max_element, minmax_element, next_permutation, prev_permutation

which do not. So in addition to not being general, it's too weak to solve the problem in question.

I found a couple of inconsistencies in the specification:
1. `set_intersection` seems to have been overlooked, it returns only an output iterator whereas the other set algorithms return tuples.
2. `set_difference` is weird; it returns a pair containing the output iterator and only one of the two input range iterators.

**Comment by ericniebler** [comment]
> I found a couple of inconsistencies in the specification:
> 1. `set_intersection` seems to have been overlooked, it returns only an output iterator whereas the other set algorithms return tuples.
> 2. `set_difference` is weird; it returns a pair containing the output iterator and only one of the two input range iterators.

There's a method to the madness. `set_intersection` stops when *either* input sequence runs out, so it does not necessarily find the end of either. `set_difference` is asymmetric; it necessarily finds the end of the first input sequence, but not the second.

**Comment by tomaszkam** [comment]
> `InputRange` will consequently become move-only as well. The same arguments about invalidation apply to output iterators, so `OutputIterator` and `OutputRange` (if/when we decide we want to support e.g. 4-legged copy) should also become move-only. I'm not arguing for or against, just pointing out that this is a sizable change.

You comment about move-only input sentinels lead me to realization of the consequence of making any iterator move-only that I was not fully aware before: you will no longer be able to use two copies of single iterator to denote an empty range. I consider this looks to be a show-stopper for this change. as it against conceptual design of iterators.

**Comment by CaseyCarter** [comment]
Addressing the original topic of the thread: yes, Semiregular must require both copy-independence and move-independence to be sane. This is a natural extension of the current requirement that Semiregular types not be volatile. I think both requirements are actually consequences of a higher-level requirement which is actually what we should specify: that the value of an object must only change through the direct application of modifier operations to that object. That requirement rules out all sorts of silliness that an algorithm can't be expected to cope with: spontaneous changes to volatile objects, copy/move dependence, simultaneous modification of values in another thread, even aliasing of references or iterators.

**Comment by ericniebler** [comment]
Yes, `Semiregular` should require copy independence.

**Comment by asutton** [comment]
I've been putting off weighing in on this because I haven't been sure
what to think about it. But after reflecting, it's not right. Copy or
move independence should never be mentioned in the requirements of a
concept.

For starters, doing so would make istream_iterators fail to be
Semiregular. I'm guessing this would fail that test:

istream_iterator<char> iter1(cin);
istream_iterator<char> iter2 = iter1;
++iter1; // Oops... also modifies iter2

But that's okay; our existing concepts for input iterators give us
guidance for how to define algorithms in a way where we don't invoke
undefined behavior.

We typically want the postconditions of an algorithm to determine the
preconditions. This is why, for example, we can't use predicates for
find() that modify elements (or the input range, really) since we may
not be able to guarantee the postcondition.

We should never have to list anything more than the postconditions of
an algorithm, and you have to consider those on a case-by-case basis.

Andrew Sutton

**Comment by ericniebler** [comment]
> istream_iterator<char> iter1(cin);
> istream_iterator<char> iter2 = iter1;
> ++iter1; // Oops... also modifies iter2

No, `++iter1` does not modify `iter2`. It invalidates `iter2`'s value, but the value is unchanged. Read [@CaseyCarter's comment](https://github.com/ericniebler/stl2/issues/38#issuecomment-108530761). Input iterators *do* have copy independence, when considering the value of the iterator, not the validity of the iterator or the value of the object the iterator points to.

**Comment by asutton** [comment]
> istream_iterator iter1(cin);
> istream_iterator iter2 = iter1;
> ++iter1; // Oops... also modifies iter2
>
> No, ++iter1 does not modify iter2. It invalidates iter2's value, but the
> value is unchanged. Read @CaseyCarter's comment. Input iterators do have
> copy independence, when considering the value of the iterator, not the
> validity of the iterator or the value of the object the iterator points to.

That's an implementation detail. Write in terms of operator* and you
still get the fact that you have copies modifying each other's state.

Copy independence is still the wrong property.

**Comment by ericniebler** [comment]
```c++
int *p = new int(42);
int *q = p;
// p and q are independent, both valid
delete q;
// now p is invalid. It's value is unmodified, though
```

Surely pointers have copy independence. I see no difference between this case and input iterators.

**Comment by asutton** [comment]
> int *p = new int(42);
> int *q = p;
> // p and q are independent, both valid
> delete q;
> // now p is invalid. It's value is unmodified, though

p is perfectly valid in this program because you aren't dereferencing
it. However, there is no object pointed at because it was just
deleted. But there's nothing about this program that tells me that any
of that matters.

> Surely pointers have copy independence. I see no difference between this
> case and input iterators.

That's not the point.

My point is that we should nowhere in the spec require copy
independence. It is not an essential property of the algorithms being
constrained. Only the postconditions matter, and the only way you
guarantee that postconditions follow from a definition is that
equality is preserved in all operations required to preserve equality.

Andrew

**Comment by ericniebler** [comment]
> p is perfectly valid in this program because you aren't dereferencing it.

[[basic.stc.dynamic.deallocation]/4:
"If the argument given to a deallocation function in the standard library is a pointer that is not the null pointer value (4.10), the deallocation function shall deallocate the storage referenced by the pointer, **rendering invalid all pointers referring to any part of the deallocated storage**."

> My point is that we should nowhere in the spec require copy independence. It is not an essential property of the algorithms being constrained.

It's an essential property of Regular types. This is set out in [_Fundamentals of Generic Programming_](http://www.stepanovpapers.com/DeSt98.pdf) (Dehrnert, Stepanov), in the section describing the copy/assign semantics of Regular types:

"Our next axiom says that we intend the C++ copy semantics:

(3) `T a = c; T b = c; a = d; assert(b==c);`

Here, after assigning the same value c to both a and b, we expect to be able to
modify a without changing the value of b. In fact, we want an even stronger
condition. If `zap` is an operation which always changes the value of its operand, we
expect the following to hold:

(4) `T a = c; T b = c; zap(a); assert(b==c && a!=b);`

That is, b and c do not continue to be equal simply because their values were
changed along with a’s, but rather because changing a’s value did not change theirs."

**Comment by asutton** [comment]
> (4) T a = c; T b = c; zap(a); assert(b==c && a!=b);
>
> That is, b and c do not continue to be equal simply because their values were
> changed along with a’s, but rather because changing a’s value did not change theirs."

I don't believe that this has appeared in any description of the
semantics of "regular" since that paper was published. I think that
EoP refines that notion to only require equality preservation for
operations that need it -- and that's certainly the approach we
adopted in n3351. There was no notion of copy independence.

And now show that this is a necessary and non-redundant requirement
(i.e., cannot be derived from postconditions) for every algorithm and
data structure in the STL without invalidating existing uses (that
aren't "weird").

I'm arguing for you to *not* make additional changes. I think
postconditions of the existing library are sufficient to derive that
requirement when needed. And where extra conditions are required, I
believe they are adequately stated in the text.

Andrew

**Comment by sean-parent** [comment]
Copy independence is only required to hold for regular procedures -

If we went back to the 98 paper, the zap procedure is an example of an
action in EoP terms - for any regular action there is an equivalent regular
function. That is  a(x) is equivalent to  x = f(x) for some f(), ignoring
performance. So we see that a regular action can always be defined in terms
of object modifications through assignment and therefore we only need to
ensure that copy independence is obeyed by assignment, no need to bring a
zap() function into the discussion.

operator++ on input iterators is an odd beast. As Andrew correctly points
out, operator++ doesn't change the value of other iterators. It invalidates
those iterators. It make it not possible to compare prior copies with
equality. It isn't transformable to a pure function (it clearly has a side
effect). We can't even say it is not a regular action as there is no way to
pass equal values to it.

input iterators are fully regular.

Sean

**Comment by ericniebler** [comment]
> operator++ on input iterators is an odd beast. As Andrew correctly points
> out, operator++ doesn't change the value of other iterators. It invalidates
> those iterators.

Actually, @CaseyCarter and I were saying that.

@sean-parent You seem to be saying that Semiregular *should* require copy independence for assignment (and that input iterators satisfy that requirement). Did I get that right? Andrew thinks differently, and I'm now confused.

**Comment by tomaszkam** [comment]
If the copy-independence is not required as the part of the `Semiregular` concept, which of the requirement placed on the `sort` function excludes `vector<proxy>` from being passed to this function. Lets assume that the `proxy` is smart reference similar to `reference_wrapper`, but instead of being rebind on assigment, it assings new value to referenced object. The same would apply for the `swap` implementation - the values of referenced objects are swapped.

I belive that `vector<proxy>` will not handled correctly by [_range-v3_ implementation of `sort`](https://github.com/ericniebler/range-v3/blob/master/include/range/v3/algorithm/sort.hpp). For example `linear_insert` function assums that after storing the element pointed by `end` iterator in `val`, the value of the copy (`val`) will not changed after modification of original object (`*end`), which is consequence of `move_backward` invocation.

**Comment by sean-parent** [comment]
Sorry for the incorrect attribution.

If by copy independence you mean through assignment then yes, this should
be required of semiregular - and input iterators are fully regular (they
have to be equality comparable, which is the only difference between
semiregular and regular).

[ As for the discussion on proxies, what most people mean by a proxy is a
user defined reference type. Such a type is not possible to create with the
current language because it breaks type deduction. Fix that issue, don't
try and weaken concepts to allow them because they still won't work. ]

Sean

**Comment by CaseyCarter** [comment]
Copy independence is not a feature specific to regular types in EoP - it's a fundamental feature of all object types. §1.3 pg. 5:
> Two objects of the same type are *equal* if and only if their states are equal. If two objects are equal, we say that one is a *copy* of the other. **Making a change to an object does not affect any copy of it.**

**Comment by tomaszkam** [comment]
> [ As for the discussion on proxies, what most people mean by a proxy is a user defined reference type. Such a type is not possible to create with the current language because it breaks type deduction. Fix that issue, don't try and weaken concepts to allow them because they still won't work. ]

My point was not about weakening concept, rather the fact that I am pointing out that I am unable to find anything in the wording that would say that types that does not expose copy-independence for assignment and copy construction (like `proxy`) are prohibited to be used with STL algorithms. This leads to my original question - should we require copy-independence and on which level?

**Comment by asutton** [comment]
If I have a lambda expression that captures an object by reference and
modifies that object in the function call operator (say, incrementing the
number of times called), would that lambda type be considered to exhibit
the property of "copy independence"?



Andrew Sutton

**Comment by ericniebler** [comment]
Lambdas are not equality comparable. A better example would be the result of calling `std::tie`. A tuple of references does *not* have copy independence AFAICT, since assigning to one writes through the references.

**Comment by asutton** [comment]
Neither are function objects, and yet both lambdas and function objects are
expected to be Semiregular.


Andrew Sutton

**Comment by ericniebler** [comment]
> Neither are function objects, and yet both lambdas and function objects are expected to be Semiregular.

Who expects that? In what context?

**Comment by asutton** [comment]
I thought that functions in n3351 and the ranges TS  were semiregular,
but apparently only copy constructible and destructible. Apparently I
was mistaken.

I want to see concrete examples where postconditions are broken when we don't
have copy independence. I was trying to craft one, but it's not
working out so well.

Andrew Sutton

**Comment by tomaszkam** [comment]
> I want to see concrete examples where postconditions are broken when we don't have copy independence. I was trying to craft one, but it's not working out so well.

I just presented the one regarding `sort` algorithm and array of `proxy` a few posts before.

**Comment by asutton** [comment]
Show an actual program that breaks. Not just names of types.


Andrew Sutton

**Comment by sean-parent** [comment]
Correct - in EoP "regular" comes from the definition of a type.

**Comment by sean-parent** [comment]
Copy independence for assignment should be required.

**Comment by ericniebler** [comment]
> I want to see concrete examples where postconditions are broken when we don't
> have copy independence. I was trying to craft one, but it's not
> working out so well.

I don't think Quick_sort would work so well when swapping unrelated objects could stomp the pivot. Is that what you're looking for?

**Comment by sean-parent** [comment]
If it is copyable in any form then copies must be independent.

**Comment by sean-parent** [comment]
The most trivial example is just swap() - it doesn't work if copies are not
independent.

**Comment by tomaszkam** [comment]
> Show an actual program that breaks. Not just names of types.

Just take a look at http://melpon.org/wandbox/permlink/dkJiEZtQwLOkOlLc. I believe that `proxy<int>*` meets all requirement stated for the `sort` algorithm, but the post condition are not held - the range is not sorted.

To be clear, I am not saying that such class should be expected to work with the algorithm. I am just expecting that I would be able to point out the requirement in the wording that will no be fulfilled by this class and makes behavior of linked undefined/unspecified.

**Comment by sean-parent** [comment]
I believe the proxy example is a good example because people often attempt
to write such classes and they are doomed to failure. The problem is that
the intention is that proxy<int> is expected to behave as an int&, but
fails. The root cause of the failure is that (T& x) when passed a proxy
will not deduce T to be an int, but to be a proxy<int>. The reason we can
swap int& in a meaningful way but not proxy<int> (through std::swap) is
because the type being swapped is deduced to int for int& but proxy<int>
for proxy<int>. Providing your own swap only fixes swap, but not other
permutations. There is no way to "patch" a proxy up to work correctly with
the current language rules.

It should be clear from the documentation (at least) that proxy<int> is not
a valid type to pass to std::swap. Another way to look at the violation
with proxy is that copy and assignment are not consistent - proxy x{y} is
not equivalent to proxy x; x = y; Hmm, does copy/assignment consistency
(along with the requirement that copies are equal) imply copy independence?
That may be true, I haven't considered it before.

**Comment by tomaszkam** [comment]
> Hmm, does copy/assignment consistency (along with the requirement that copies are equal) imply copy independence? That may be true, I haven't considered it before.

I belive that this is not true. Notice that for my `proxy<int>` example, for object created using default constructor (there are equal, so they are copies) the copy-independence is not held, regardless of the semantics of copy/move constructor. We do not need to use copy/assignment to create copies.

**Comment by sean-parent** [comment]
On Mon, Jun 8, 2015 at 2:38 PM, tomaszkam <notifications@github.com> wrote:

> Hmm, does copy/assignment consistency (along with the requirement that
> copies are equal) imply copy independence? That may be true, I haven't
> considered it before.
>
> I belive that this is not true. Notice that for my proxy<int> example,
> for object created using default constructor (there are equal, so they are
> copies) the copy-independence is not held, regardless of the semantics of
> copy/move constructor. We do not need to use copy/assignment to create
> copies.
>
Excellent point. I agree.

Sean

**Comment by ericniebler** [comment]
@sean-parent writes:
> There is no way to "patch" a proxy up to work correctly with the current language rules.

And I'm not (yet) convinced any sane set of language rules can help here. There are two objects: the proxy and the referent. I suspect there are times when you want the deduced type to be `int` and other times you want to get `proxy<int>`. I'm not sure how a language feature can/should deal with this complexity. At least with a library solution, we can use a named function when we want the "other" behavior.

**Comment by ericniebler** [comment]
@CaseyCarter, copy independence should be required by `Copyable`, but the latest draft doesn't say that. Are you doing adding it or shall I?

**Comment by CaseyCarter** [comment]
One of the elegant consequences of the as-yet-unworded requirement that the value of equality-preserving expressions must be stable in the absence of direct modifications of the input objects is that copy independence magically appears. CopyConstructible requires `T{t}` to be equality-preserving and that `T{t}` is equal to `t`. When those requirements are combined with a modification - such as Copyable's assignment operator - you have copy independence. E.g. if `t` of type `T` is `Copyable` then `auto u = t; u = /* something */;` does not directly modify `t` so `T{t}` must still be equal to the value that `t` had beforehand.

**Comment by ericniebler** [comment]
My understanding about the the as-yet-unworded requirement of the value stability of equality-preserving expressions related only to those expressions appearing in the algorithms. E.g. I can pass two `reference_wrapper`s to `std::swap` so long as they do not alias each other. But copy-independence is a property of the *type*. `reference_wrapper` does not have copy-independence in this way.

**Comment by ericniebler** [comment]
> E.g. I can pass two reference_wrappers to std::swap so long as they do not alias each other.

Well, you get the idea. Ahem.

**Comment by ericniebler** [comment]
No wait, you're right. I get it.
